README
NAME
    AnyEvent::XMPP - An implementation of the XMPP Protocol

VERSION
    Version 0.51

SYNOPSIS
       use AnyEvent::XMPP::Connection;

    or:

       use AnyEvent::XMPP::IM::Connection;

    or:

       use AnyEvent::XMPP::Client;

DESCRIPTION
    This is the head module of the AnyEvent::XMPP XMPP client protocol (as
    described in RFC 3920 and RFC 3921) framework.

    AnyEvent::XMPP::Connection is a RFC 3920 conforming "XML" stream
    implementation for clients, which handles TCP connect up to the resource
    binding. And provides low level access to the XML nodes on the XML
    stream along with some high level methods to send the predefined XML
    stanzas.

    AnyEvent::XMPP::IM::Connection is a more high level module, which is
    derived from AnyEvent::XMPP::Connection. It handles all the instant
    messaging client functionality described in RFC 3921.

    AnyEvent::XMPP::Client is a multi account client class. It manages
    connections to multiple XMPP accounts and tries to offer a nice high
    level interface to XMPP communication.

    For a list of "Supported extensions" see below.

    There are also other modules in this distribution, for example:
    AnyEvent::XMPP::Util, AnyEvent::XMPP::Writer, AnyEvent::XMPP::Parser and
    those I forgot :-) Those modules might be helpful and/or required if you
    want to use this framework for XMPP.

    See also AnyEvent::XMPP::Writer for a discussion about the brokenness of
    XML in the XMPP specification.

    If you have any questions or seek for help look below under "SUPPORT".

REQUIREMENTS
    One of the major drawbacks I see for AnyEvent::XMPP is the long list of
    required modules to make it work.

    AnyEvent
        For the I/O events, timers, TCP, TLS, DNS and I/O buffering.

    Object::Event
        The former AnyEvent::XMPP::Event module has been outsourced to the
        Object::Event module to provide a more generic way for more other
        modules to register and call event callbacks.

    XML::Writer
        For writing "XML".

    XML::Parser::Expat
        For parsing partial "XML" stuff.

    MIME::Base64
        For SASL authentication

    Authen::SASL
        For SASL authentication

    Net::LibIDN
        For stringprep profiles to handle JIDs.

    Digest::SHA1
        For component authentication and old-style authentication.

    And yes, all these are essential for XMPP communication. Even though
    'instant messaging' and 'presence' is a quite simple problem XMPP
    somehow was successful at making the task complicated enough to keep me
    busy for a long time. But all of that time wasn't only for the
    technology required to get it started, mostly it was for all the quirks,
    hacks and badly applied "XML" in the protocol which complicated the
    matter.

RELEASE NOTES
    Here are some notes to the last releases (release of this version is at
    top):

  Version
    *   0.51

        Maintenance release. Added a patch which fixes Object::Event
        compatibility and another fix w.r.t. memory leak in the parser. And
        added the original node to AnyEvent::XMPP::IM::Message (thanks go to
        mons@cpan.org).

        NOTE: Version 0.6 of AnyEvent::XMPP will be API incompatible! If you
        are already eager to try the new version out contact me!

    *   0.5

        Maintenance release. Added a patch from Marcus Dubois for
        Ext::Pubsub. Also fixed some memleaks in AnyEvent::XMPP::Parser.

        Also wanted to note that the next version of AnyEvent::XMPP will
        have an incompatible API. If you are eager to try out the new
        complete rewrite of AnyEvent::XMPP contact me.

    *   0.4

        Minor fixes and feature enhancements: Added old_style_ssl option for
        direct port 5223 SSL connections. Providing 'get_own_contact' for
        keeping track of own resources.

        The AnyEvent::XMPP::Ext::MUC extension was rewritten and provides a
        more sane API now.

        For details consult the Changes file in the distribution.

    *   0.3

        Fixed some small bugs and improved documentation a bit, especially
        w.r.t. parameter passing of host and ports.

    *   0.2

        Renamed module from Net::XMPP2 to AnyEvent::XMPP. Net::XMPP2 is
        herby deprecated!

        Rewrote the low-level socket stuff to use AnyEvent::Socket and
        AnyEvent::Handle. Removed blocking write functionality, which can't
        be supported that easily with AnyEvent::Handle (however, if you want
        to wait until the send-buffer is empty you best use the
        "send_buffer_empty" event of AnyEvent::XMPP::Connection).

        For more details consult the Changes file of the AnyEvent::XMPP
        distribution.

    *   older

        For older release notes please have a look at the Changes file or
        CPAN.

  TODO
    There are still lots of items on the TODO list (see also the TODO file
    in the distribution of AnyEvent::XMPP).

TEST SUITE
    If you are a developer and want to test either a server or maybe just
    whether this module passes some basic tests you might want to run the
    developer test suite.

    This test suite is not enabled by default because it requires some human
    interaction to set it up, please see AnyEvent::XMPP::TestClient for
    hints about the setup procedure for the test suite.

    I wrote the test suite mostly because I wanted to make sure I didn't
    break something essential before a release. The tests don't cover
    everything and I don't plan to write a test for every single function in
    the API, that would slow down development considerably for me. But I
    hope that some grave show stopper bugs in releases are prevented with
    this test suite.

    The tests are also useful if you want to test a server implementation.
    But there are maybe of course conformance issues with AnyEvent::XMPP
    itself, so if you find something where AnyEvent::XMPP doesn't conform to
    the XMPP RFCs or XEPs consult the BUGS section below.

    If you find a server that doesn't handle something correctly but you
    need to interact with it you are free to implement workarounds and send
    me a patch, or even ask me whether I might want to look into the issue
    (I can't guarantee anything here, but I want this module to be as
    interoperable as possible. But if the implementation of a workaround for
    some non-conformant software will complicate the code too much I'm
    probably not going to implement it.).

    Of course, if you find a bug in some server implementation don't forget
    to file a bugreport to them, one hack less in AnyEvent::XMPP means more
    time for bug fixing and improvements and new features.

Why (yet) another XMPP module?
    The main outstanding feature of this module in comparison to the other
    XMPP (aka Jabber) modules out there is the support for AnyEvent.
    AnyEvent permits you to use this module together with other I/O event
    based programs and libraries (ie. Gtk2 or Event).

    The other modules could often only be integrated in those applications
    or libraries by using threads. I decided to write this module because I
    think CPAN lacks an event based XMPP module. Threads are unfortunately
    not an alternative in Perl at the moment due the limited threading
    functionality they provide and the global speed hit. I also think that a
    simple event based I/O framework might be a bit easier to handle than
    threads.

    Another thing was that I didn't like the APIs of the other modules. In
    AnyEvent::XMPP I try to provide low level modules for speaking XMPP as
    defined in RFC 3920 and RFC 3921 (see also AnyEvent::XMPP::Connection
    and AnyEvent::XMPP::IM::Connection). But I also try to provide a high
    level API for easier usage for instant messaging tasks and clients (eg.
    AnyEvent::XMPP::Client).

Supported extensions
    See AnyEvent::XMPP::Ext for a list.

EXAMPLES
    Following examples are included in this distribution:

    samples/simple_example_1
        This example script just connects to a server and sends a message
        and also displays incoming messages on stdout.

    samples/conference_lister
        See below.

    samples/room_lister
        See below.

    samples/room_lister_stat
        These three scripts implements a global room scan.
        "conference_lister" takes a list of servers (the file is called
        "servers.xml" which has the same format as the xml file at
        <http://www.jabber.org/servers.xml>). It then scans all servers for
        chat room services and lists them into a file "conferences.stor",
        which is a Storable dump.

        "room_lister" then reads that file and queries all services for
        rooms, and then all rooms for their occupants. The output file is
        "room_data.stor", also a Storable dump, which in turn can be read
        with "room_lister_stat", which transform the data structures into
        something human readable.

        These scripts are a bit hacky and quite complicated, but maybe it's
        of any value for someone. You might note "EVQ.pm" in samples which
        is a module that handles request-throttling (You don't want to flood
        the server and risk getting the admins attention :).

    samples/simple_component
        This is a (basic) skeleton for a jabber component.

    samples/simple_oob_retriever
        This is a simple out of band file transfer receiver bot. It uses
        "curl" to fetch the files and also has the sample functionality of
        sending a file url for someone who sends the bot a 'send <filename>'
        message.

    samples/simple_register_example
        This is a example script which allows you to register, unregister
        and change your password for accounts. Execute it without arguments
        for more details.

    samples/disco_info
        This is a small example tool that allows you to fetch the software
        version, disco info and disco items information about a JID.

    samples/talkbot
        This is a simple bot that will read lines from a file and recite
        them when you send it a message. It will also automatically allow
        you to subscribe to it. Start it without commandline arguments to be
        informed about the usage.

    samples/retrieve_roster
        This is a simple example script that will retrieve the roster for an
        account and print it to stdout. You start it like this:

           samples/# ./retrieve_roster <jid> <password>

    samples/display_avatar
        This is just a small example which should display the avatar of the
        account you connect to. It can be used like this:

           samples/# ./display_avatar <jid> <password>

    For others, which the author might forgot or didn't want to list here
    see the "samples/" directory.

    More examples will be included in later releases, please feel free to
    ask the "AUTHOR" if you have any questions about the API. There is also
    an IRC channel, see "SUPPORT".

AUTHOR
    Robin Redeker, "<elmex at ta-sa.org>", JID: "<elmex at jabber.org>"

BUGS
    Please note that I'm currently (July 2007) the only developer on this
    project and I'm very busy with my studies in Computer Science. If you
    want to ease my workload or want timely releases, please send me patches
    instead of bug reports or feature requests. I won't forget the reports
    or requests if you can't or didn't send patches, but I can't gurantee
    immediate response. But I will of course try to fix/implement them as
    soon as possible!

    Also try to be as precise as possible with bug reports, if you can't
    send a patch, it would be best if you find out which code doesn't work
    and tell me why.

    Please report any bugs or feature requests to "bug-net-xmpp2 at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=AnyEvent-XMPP>. I will
    be notified and then you'll automatically be notified of progress on
    your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc AnyEvent::XMPP

    You can also look for information at:

    *   IRC: AnyEvent::XMPP IRC Channel

          IRC Network: http://freenode.net/
          Server     : chat.freenode.net
          Channel    : #ae_xmpp

          Feel free to join and ask questions!

    *   AnyEvent::XMPP Project Site

        <http://www.ta-sa.org/net_xmpp2>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/AnyEvent-XMPP>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/AnyEvent-XMPP>

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=AnyEvent-XMPP>

    *   Search CPAN

        <http://search.cpan.org/dist/AnyEvent-XMPP>

ACKNOWLEDGEMENTS
    Thanks to the XSF for the development of an open instant messaging
    protocol (even though it uses "XML").

    And thanks to all people who had to listen to my desperate curses about
    the brokenness/braindeadness of XMPP. Without you I would've never
    brought this module to a usable state.

    Thanks to:

    *   J. Cameijo Cerdeira

        For pointing out a serious bug in "split_jid" in
        AnyEvent::XMPP::Util and suggesting to add a timeout argument to the
        "connect" method of AnyEvent::XMPP::SimpleConnection.

    *   Carlo von Loesch (aka lynX) <http://www.psyced.org/>

        For pointing out some typos.

    *   All other people ..

        ... I mentioned in the CONTRIBUTORS file which comes with the
        AnyEvent::XMPP distribution.

COPYRIGHT & LICENSE
    Copyright 2007, 2008 Robin Redeker, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for AnyEvent::XMPP:


0.51     Fri Aug 14 10:53:23 CEST 2009

   !!!!!!!!!!!!!!!
   !!! NOTE: The next version (0.6) of AnyEvent::XMPP will have an incompatible API!
   !!!!!!!!!!!!!!!

         - small uncleanliness fixed in ::Parser.
         - thanks to mons@cpan.org, who found an incompat.
           between OE 1.1 and AnyEvent::XMPP (thanks to mons@cpan.org).
         - added message id and xml_node to IM::Message/Delayed
           (thanks to mons@cpan.org).


0.5 Mon Mar 16 21:13:13 CET 2009

   !!!!!!!!!!!!!!!
   !!! NOTE: The next version of AnyEvent::XMPP will have an incompatible API!
   !!!!!!!!!!!!!!!

         - added patch for Ext::Pubsub.
         - melo fixed some memleaks.
         - fixed another long time memleak in the parser.

0.4 Sun Feb 15 22:59:38 CET 2009
         - added a $dont_retrieve_vcard to ::Ext::VCard, to
           prevent retrieval when session_ready event comes.
         - implemented support for old-style (tcp port 5223) SSL connections,
           use the old_style_ssl option for ::Connection to enable TLS handshaking
           upon TCP connect.
         - added $message argument to subscription related events. the contents
           comes from the <status> elements in the presence stanza.
         - made special contact for keeping track of the presences of our own
           resources. See ::IM::Roster::get_own_contact() and ::IM::Contact::is_me().
         - minor fixes in ::Ext::MUC.
         - added 'stream_version_override' parameter to ::Connection, mostly
           for testing purposes with ejabberd (to enforce old-jabber-style
           authentication).
         - partially reversed a patch w.r.t. hostname usage for SASL mechanism.
           A flag 'use_host_as_sasl_hostname' is provided for giving the hostname
           to SASL.
         - rewrote AnyEvent::XMPP::Ext::MUC a bit and made the API more sane.
           for example: join_room doesn't take a callback anymore.
         - added more C<disco_feature> methods to the extensions, so that
           they can be registered to an ::Ext::Disco extension to display
           support for certain features.
         - xmpp_datetime_as_timestamp fixed w.r.t. UTC based timestamps.
         - added possibility to override iq_xml events, and documented
           how to overtake handling of message, presence and iq stanzas.
         - added patch from Pedro Melo to give stanzas sent by a ::Component
           the right namespace.
         - added patch from Pedro Melo to fix passing of the host in ::Component.

0.3 Mon Jan 28 23:00:58 CET 2009
         - fixed a bug in disconnect error reporting.
         - fixed bad parameter passing of AnyEvent::XMPP::Client to
           AnyEvent::XMPP::IM::Account, also 'host' is used for overriding
           the destination host now (and not 'domain', as it was in version
           0.2).
         - half fixed RT #42194 (Domain vs. host and GSSAPI),
           passing the hostname specified by 'host' for ::Connection to
           Authen::SASL's client_new method.
         - fixed bug in ::XMPP::Client where remove_account didn't work
           as expected (and set_accounts).

0.2 Sun Dec 28 15:09:25 CET 2008
         - Using the better tested AnyEvent::Handle and
           AnyEvent::Socket for DNS resolution, TCP connect and TLS now.
         - override_host and override_port have been removed.
         - weird DNS SRV handling and comments have been removed.
         - removed blocked_write option and drain function, send_buffer_empty event
           should be used. blocking behaviour should be emulated with send_buffer_empty
           at the toplevel of the program.
         - improved message tracking api in ::IM::Account
         - added added_account/removed_account events to ::Client
         - added set_accounts() function to ::Client
         - removed unneccessary sample (devcl)
         - renamed module from Net::XMPP2 to AnyEvent::XMPP
         - minor additions to ::Ext::Pubsub.

0.14 Fri Aug  1 12:13:11 CEST 2008
         - NOTICE: This is the last release of this module under the name
                   Net::XMPP2. All further releases will be done under the name
                   AnyEvent::XMPP!
         - applied fixes from git://github.com/melo/net--xmpp2.git
           branch component-reply-with-from.
         - fixed small bug in simple_component.
         - fixed a debugging bug in the drain method of SimpleConnection,
           where the written data wasn't passed to the debugging events.
         - improved the documentation a bit and fixed typos.

0.13 Tue Apr 15 13:07:02 CEST 2008
         - added a 'send_buffer_empty' event, to be able to wait
           for an empty send buffer. and also the C<drain> method, which
           will block until the send buffer is empty. also added a C<blocking_write>
           option for L<Net::XMPP2::Connection>.
         - added a default handler to L<Net::XMPP2::Parser> which
           recognizes and stores CDATA now correctly, so that it can
           be retrieved from the nodes structure by the C<as_string> method
           of L<Net::XMPP2::Node> correctly.

0.12 Fri Apr 11 16:30:40 CEST 2008
         - API CHANGE: The C<connect> method of L<Net::XMPP2::SimpleConnection>
           and thus L<Net::XMPP2::Connection> is now non-blocking!
           IMPORTANT: Please revisit your connection code and adjust accordingly!
         - improved the documentation w.r.t. event callback arguments a bit
         - improved the documentation of the send_iq method of Net::XMPP2::Connection
         - added samples/display_my_avatar
         - added support for XEP-0054 (vCard) and added a test for it (z_07_vcard.t)
         - replaced the big and unmaintainable MUC test by some smaller
           tests.
         - fixed a bug with disconnect event propagation by Net::XMPP2::Client
         - added a connect_timeout argument to Net::XMPP2::Connection and

lib/AnyEvent/XMPP.pm
package AnyEvent::XMPP;
no warnings;
use strict;

=head1 NAME

AnyEvent::XMPP - An implementation of the XMPP Protocol

=head1 VERSION

Version 0.51

=cut

our $VERSION = '0.51';

=head1 SYNOPSIS

   use AnyEvent::XMPP::Connection;

or:

   use AnyEvent::XMPP::IM::Connection;

or:

   use AnyEvent::XMPP::Client;

=head1 DESCRIPTION

This is the head module of the L<AnyEvent::XMPP> XMPP client protocol (as described in
RFC 3920 and RFC 3921) framework.

L<AnyEvent::XMPP::Connection> is a RFC 3920 conforming "XML" stream implementation
for clients, which handles TCP connect up to the resource binding. And provides
low level access to the XML nodes on the XML stream along with some high
level methods to send the predefined XML stanzas.

L<AnyEvent::XMPP::IM::Connection> is a more high level module, which is derived
from L<AnyEvent::XMPP::Connection>. It handles all the instant messaging client
functionality described in RFC 3921.

L<AnyEvent::XMPP::Client> is a multi account client class. It manages connections
to multiple XMPP accounts and tries to offer a nice high level interface
to XMPP communication.

For a list of L</Supported extensions> see below.

There are also other modules in this distribution, for example:
L<AnyEvent::XMPP::Util>, L<AnyEvent::XMPP::Writer>, L<AnyEvent::XMPP::Parser> and those I
forgot :-) Those modules might be helpful and/or required if you want to use
this framework for XMPP.

See also L<AnyEvent::XMPP::Writer> for a discussion about the brokenness of XML in the XMPP
specification.

If you have any questions or seek for help look below under L</SUPPORT>.

=head1 REQUIREMENTS

One of the major drawbacks I see for AnyEvent::XMPP is the long list of required
modules to make it work.

=over 4

=item L<AnyEvent>

For the I/O events, timers, TCP, TLS, DNS and I/O buffering.

=item L<Object::Event>

The former L<AnyEvent::XMPP::Event> module has been outsourced to the L<Object::Event>
module to provide a more generic way for more other modules to register and call
event callbacks.

=item L<XML::Writer>

For writing "XML".

=item L<XML::Parser::Expat>

For parsing partial "XML" stuff.

=item L<MIME::Base64>

For SASL authentication

=item L<Authen::SASL>

For SASL authentication

=item L<Net::LibIDN>

For stringprep profiles to handle JIDs.

=item L<Digest::SHA1>

For component authentication and old-style authentication.

=back

And yes, all these are essential for XMPP communication. Even though 'instant
messaging' and 'presence' is a quite simple problem XMPP somehow was successful
at making the task complicated enough to keep me busy for a long time.  But all
of that time wasn't only for the technology required to get it started, mostly
it was for all the quirks, hacks and badly applied "XML" in the protocol which
complicated the matter.

=head1 RELEASE NOTES

Here are some notes to the last releases (release of this version is at top):

=head2 Version

=over 4

=item * 0.51

Maintenance release. Added a patch which fixes L<Object::Event> compatibility
and another fix w.r.t. memory leak in the parser. And added the original node
to L<AnyEvent::XMPP::IM::Message> (thanks go to mons@cpan.org).

B<NOTE:> Version 0.6 of L<AnyEvent::XMPP> will be API incompatible! If you
are already eager to try the new version out contact me!

=item * 0.5

Maintenance release. Added a patch from Marcus Dubois for Ext::Pubsub.
Also fixed some memleaks in L<AnyEvent::XMPP::Parser>.

Also wanted to note that the next version of AnyEvent::XMPP will have an
incompatible API. If you are eager to try out the new complete rewrite of
AnyEvent::XMPP contact me.

=item * 0.4

Minor fixes and feature enhancements: Added old_style_ssl option for direct
port 5223 SSL connections. Providing 'get_own_contact' for keeping
track of own resources.

The L<AnyEvent::XMPP::Ext::MUC> extension was rewritten and provides a more
sane API now.

For details consult the Changes file in the distribution.

=item * 0.3

Fixed some small bugs and improved documentation a bit, especially w.r.t.
parameter passing of host and ports.

=item * 0.2

Renamed module from L<Net::XMPP2> to L<AnyEvent::XMPP>. L<Net::XMPP2> is herby
deprecated!

Rewrote the low-level socket stuff to use L<AnyEvent::Socket> and L<AnyEvent::Handle>.
Removed blocking write functionality, which can't be supported that
easily with L<AnyEvent::Handle> (however, if you want to wait until the send-buffer
is empty you best use the C<send_buffer_empty> event of L<AnyEvent::XMPP::Connection>).

For more details consult the Changes file of the AnyEvent::XMPP distribution.

=item * older

For older release notes please have a look at the Changes file or CPAN.

=back

=head2 TODO

There are still lots of items on the TODO list (see also the TODO file
in the distribution of AnyEvent::XMPP).

=head1 TEST SUITE

If you are a developer and want to test either a server or maybe just whether
this module passes some basic tests you might want to run the developer test
suite.

This test suite is not enabled by default because it requires some human
interaction to set it up, please see L<AnyEvent::XMPP::TestClient> for hints about
the setup procedure for the test suite.

I wrote the test suite mostly because I wanted to make sure I didn't break
something essential before a release. The tests don't cover everything and I
don't plan to write a test for every single function in the API, that would
slow down development considerably for me. But I hope that some grave show
stopper bugs in releases are prevented with this test suite.

The tests are also useful if you want to test a server implementation. But
there are maybe of course conformance issues with L<AnyEvent::XMPP> itself, so if
you find something where L<AnyEvent::XMPP> doesn't conform to the XMPP RFCs or XEPs
consult the L<BUGS> section below.

If you find a server that doesn't handle something correctly but you need to
interact with it you are free to implement workarounds and send me a patch, or
even ask me whether I might want to look into the issue (I can't guarantee
anything here, but I want this module to be as interoperable as possible. But
if the implementation of a workaround for some non-conformant software will
complicate the code too much I'm probably not going to implement it.).

Of course, if you find a bug in some server implementation don't forget to file
a bugreport to them, one hack less in L<AnyEvent::XMPP> means more time for bug
fixing and improvements and new features.

=head1 Why (yet) another XMPP module?

The main outstanding feature of this module in comparison to the other XMPP
(aka Jabber) modules out there is the support for L<AnyEvent>. L<AnyEvent>
permits you to use this module together with other I/O event based programs and
libraries (ie. L<Gtk2> or L<Event>).

The other modules could often only be integrated in those applications or
libraries by using threads. I decided to write this module because I think CPAN
lacks an event based XMPP module. Threads are unfortunately not an alternative
in Perl at the moment due the limited threading functionality they provide and
the global speed hit. I also think that a simple event based I/O framework
might be a bit easier to handle than threads.

Another thing was that I didn't like the APIs of the other modules. In
L<AnyEvent::XMPP> I try to provide low level modules for speaking XMPP as defined
in RFC 3920 and RFC 3921 (see also L<AnyEvent::XMPP::Connection> and
L<AnyEvent::XMPP::IM::Connection>). But I also try to provide a high level API for
easier usage for instant messaging tasks and clients (eg. L<AnyEvent::XMPP::Client>).

=head1 Supported extensions

See L<AnyEvent::XMPP::Ext> for a list.

=head1 EXAMPLES

Following examples are included in this distribution:

=over 4

=item B<samples/simple_example_1>

This example script just connects to a server and sends a message and
also displays incoming messages on stdout.

=item B<samples/conference_lister>

See below.

=item B<samples/room_lister>

See below.

=item B<samples/room_lister_stat>

These three scripts implements a global room scan.  C<conference_lister> takes
a list of servers (the file is called C<servers.xml> which has the same format as
the xml file at L<http://www.jabber.org/servers.xml>). It then scans all
servers for chat room services and lists them into a file C<conferences.stor>,
which is a L<Storable> dump.

C<room_lister> then reads that file and queries all services for rooms, and then
all rooms for their occupants. The output file is C<room_data.stor>, also a L<Storable>
dump, which in turn can be read with C<room_lister_stat>, which transform
the data structures into something human readable.

These scripts are a bit hacky and quite complicated, but maybe it's of any
value for someone. You might note L<samples/EVQ.pm> which is a module that
handles request-throttling (You don't want to flood the server and risk
getting the admins attention :).

=item B<samples/simple_component>

This is a (basic) skeleton for a jabber component.

=item B<samples/simple_oob_retriever>

This is a simple out of band file transfer receiver bot.  It uses C<curl> to
fetch the files and also has the sample functionality of sending a file url for
someone who sends the bot a 'send <filename>' message.

=item B<samples/simple_register_example>

This is a example script which allows you to register, unregister and change
your password for accounts. Execute it without arguments for more details.

=item B<samples/disco_info>

This is a small example tool that allows you to fetch the software version,
disco info and disco items information about a JID.

=item B<samples/talkbot>

This is a simple bot that will read lines from a file and recite them
when you send it a message. It will also automatically allow you to subscribe
to it. Start it without commandline arguments to be informed about the usage.

=item B<samples/retrieve_roster>

This is a simple example script that will retrieve the roster
for an account and print it to stdout. You start it like this:

   samples/# ./retrieve_roster <jid> <password>

=item B<samples/display_avatar>

This is just a small example which should display the avatar
of the account you connect to. It can be used like this:

   samples/# ./display_avatar <jid> <password>

=back

For others, which the author might forgot or didn't want to
list here see the C<samples/> directory.

More examples will be included in later releases, please feel free to ask the
L</AUTHOR> if you have any questions about the API. There is also an IRC
channel, see L</SUPPORT>.

=head1 AUTHOR

Robin Redeker, C<< <elmex at ta-sa.org> >>, JID: C<< <elmex at jabber.org> >>

=head1 BUGS

Please note that I'm currently (July 2007) the only developer on this project
and I'm very busy with my studies in Computer Science. If you want to ease my
workload or want timely releases, please send me patches instead of bug reports
or feature requests. I won't forget the reports or requests if you can't or
didn't send patches, but I can't gurantee immediate response. But I will of
course try to fix/implement them as soon as possible!

Also try to be as precise as possible with bug reports, if you can't send a
patch, it would be best if you find out which code doesn't work and tell me
why.

Please report any bugs or feature requests to
C<bug-net-xmpp2 at rt.cpan.org>, or through the web interface at
L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=AnyEvent-XMPP>.
I will be notified and then you'll automatically be notified of progress on
your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc AnyEvent::XMPP

You can also look for information at:

=over 4

=item * IRC: AnyEvent::XMPP IRC Channel

  IRC Network: http://freenode.net/
  Server     : chat.freenode.net
  Channel    : #ae_xmpp

  Feel free to join and ask questions!

=item * AnyEvent::XMPP Project Site

L<http://www.ta-sa.org/net_xmpp2>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/AnyEvent-XMPP>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/AnyEvent-XMPP>

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=AnyEvent-XMPP>

=item * Search CPAN

L<http://search.cpan.org/dist/AnyEvent-XMPP>

=back

=head1 ACKNOWLEDGEMENTS

Thanks to the XSF for the development of an open instant messaging protocol (even though it uses "XML").

And thanks to all people who had to listen to my desperate curses about the
brokenness/braindeadness of XMPP. Without you I would've never brought this
module to a usable state.

Thanks to:

=over 4

=item * J. Cameijo Cerdeira

For pointing out a serious bug in C<split_jid> in L<AnyEvent::XMPP::Util>
and suggesting to add a timeout argument to the C<connect> method of
L<AnyEvent::XMPP::SimpleConnection>.

=item * Carlo von Loesch (aka lynX) L<http://www.psyced.org/>

For pointing out some typos.

=item * All other people ..

... I mentioned in the CONTRIBUTORS file which comes with the L<AnyEvent::XMPP>
distribution.

=back

=head1 COPYRIGHT & LICENSE

Copyright 2007, 2008 Robin Redeker, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut

1; # End of AnyEvent::XMPP
